prost/src/open_enum.rs

```diff
@@ -0,0 +1,198 @@
+    /// [`unwrap_or_else`] and pass a lazily evaluated closure to it.
+    ///
+    /// [`unwrap_or_else`]: #method.unwrap_or_else
+    pub fn unwrap_or(self, default: T) -> T {
+        match self {
+            Self::Known(v) => v,
+            Self::Unknown(_) => default,
+        }
+    }
+
+    /// Returns the known value of the open enum, or, if the value is unknown,
+    /// returns the value computed from the field's raw integer value by the
+    /// provided closure.
+    pub fn unwrap_or_else<F>(self, f: F) -> T
+    where
+        F: FnOnce(i32) -> T,
+    {
+        match self {
+            Self::Known(v) => v,
+            Self::Unknown(v) => f(v),
+        }
+    }
+
+    /// Returns the known value of the open enum, or, if the value is unknown,
+    /// returns the default value of the enum type.
+    pub fn unwrap_or_default(self) -> T
+    where
+        T: Default,
+    {
+        match self {
+            Self::Known(v) => v,
+            Self::Unknown(_) => T::default(),
+        }
+    }
+
+    /// If the value of the open enum is known, returns it in `Some`, otherwise
+    /// returns `None`.
+    pub fn known(self) -> Option<T> {
+        match self {
+            Self::Known(v) => Some(v),
+            Self::Unknown(_) => None,
+        }
+    }
+
+    /// If the value of the open enum is known, returns it in `Ok`, otherwise
+    /// returns the provided error value in `Err`.
+    ///
+    /// Arguments passed to `known_or` are eagerly evaluated; if you are passing
+    /// the result of a function call, it is recommended to use
+    /// [`known_or_else`] and pass a lazily evaluated closure to it.
+    ///
+    /// [`known_or_else`]: #method.known_or_else
+    pub fn known_or<E>(self, err: E) -> Result<T, E> {
+        match self {
+            Self::Known(v) => Ok(v),
+            Self::Unknown(_) => Err(err),
+        }
+    }
+
+    /// If the value of the open enum is known, returns it in `Ok`, otherwise
+    /// returns `Err` with the value computed from the field's raw integer value
+    /// by the provided closure.
+    pub fn known_or_else<E, F>(self, err: F) -> Result<T, E>
+    where
+        F: FnOnce(i32) -> E,
+    {
+        match self {
+            Self::Known(v) => Ok(v),
+            Self::Unknown(v) => Err(err(v)),
+        }
+    }
+}
+
+impl<T> Message for OpenEnum<T>
+where
+    T: Clone + Into<i32> + Debug + Send + Sync,
+    i32: TryInto<T>,
+{
+    fn encoded_len(&self) -> usize {
+        self.to_raw().encoded_len()
+    }
+
+    fn encode_raw(&self, buf: &mut impl BufMut) {
+        self.to_raw().encode_raw(buf)
+    }
+
+    fn merge_field(
+        &mut self,
+        tag: u32,
+        wire_type: WireType,
+        buf: &mut impl Buf,
+        ctx: DecodeContext,
+    ) -> Result<(), DecodeError> {
+        let mut raw = 0;
+        <i32 as Message>::merge_field(&mut raw, tag, wire_type, buf, ctx)?;
+        *self = OpenEnum::from_raw(raw);
+        Ok(())
```

Comment:
This is the same function as `encoding::enumeration::merge`, right?

Reply:
No, `<i32 as Message>::merge_field` calls to `skip_field` if the tag argument is not 1.